[pt2] enable meta tests for `foreach` ops

isuruf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112779
* #112781
* #112778
* #112281
* __->__ #112280

and use a separate foreach_pointwise_without_values_kwarg_op_db

Co-authored-by: Isuru Fernando <ifernando@quansight.com>

lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8825,8 +8825,10 @@ def __init__(
         arity: int = 3,
         rightmost_supports_scalar: bool = False,
         rightmost_supports_scalarlist: bool = False,
+        without_values_kwarg: bool = False,
```
What's the whole deal with the values kwarg??

isuruf replied:
I'm fixing it in https://github.com/pytorch/pytorch/pull/112781 and reverted this change for this PR.

lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8825,8 +8825,10 @@ def __init__(
         rightmost_supports_scalarlist: bool = False,
+        without_values_kwarg: bool = False,
     ):
         super().__init__(3 + 1, True, True)
```
Preexisting issue, but this looks wrong? It's not forwarding the args

lezcano replied:
what about the arity?

isuruf replied:
Should the default value be `4`?

lezcano replied:
the arity is not being passed to super...

isuruf replied:
Right. It's sending `3 + 1` to super and I'm not sure what the correct value should be.

lezcano replied:
This function has an arity parameter, just forward it?

isuruf replied:
Fixed now

lezcano commented on test/test_meta.py:
```diff
@@ -57,6 +59,15 @@
+    foreach_binary_op_db +
+    foreach_pointwise_op_db +
+    foreach_reduce_op_db +
+    foreach_lerp_op_db
```
Perhaps rather than calling if `lerp` calling it `others`?

lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -8561,12 +8561,19 @@ class ForeachRightmostArgType(enum.Enum):
+    def __init__(self, *args, disable_fastpath=False, ref_args=None, **kwargs):
         super().__init__(*args, **kwargs)
-        self.ref_args = self.args
+        if ref_args:
+            self.ref_args = ref_args
+        else:
+            self.ref_args = self.args
```
Nit 
```suggestion
        self.ref_args = ref_args or self.args
```

isuruf committed:
[pt2] enable meta tests for `foreach` ops

and add explanation for ref_args

Co-authored-by: Nikita Karetnikov <nikita@karetnikov.org>

ghstack-source-id: 116618dfa35f75aed152ab464f5f394caf94dcec
Pull Request resolved: https://github.com/pytorch/pytorch/pull/112280

isuruf commented:
Rebased to fix the author commit info and had the wrong base branch. Fixed now.

isuruf commented:
@pytorchbot merge

isuruf commented:
@pytorchbot merge

isuruf commented:
I'm gonna resend the PR.
